### PR TITLE
overlord/snapstate: add timings to critical task handlers and the backend

### PR DESCRIPTION
**Pull request by stolowski**
Collect timings in some critical task handlers of snapmgr, and around select operations of LinkSnap (such as fc-cache execution).

### CODE REVIEW

**Comment by pedronis** [comment]
high-level: we should measure also around store calls in refreshhints.go and catalogrefresh.go

File: overlord/snapstate/autorefresh.go [hunk 1]
```diff
@@ -368,6 +369,9 @@ func (m *autoRefresh) refreshScheduleWithDefaultsFallback() (ts []*timeutil.Sche
 
 // launchAutoRefresh creates the auto-refresh taskset and a change for it.
 func (m *autoRefresh) launchAutoRefresh() error {
+	perfTimings := timings.New(map[string]string{"ensure": "auto-refresh"})
+	tm := perfTimings.StartSpan("auto-refresh", "prepare auto-refresh change")
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go]: I'm not sure what is the best verb here, "query store and setup auto-refresh change" ?

Comment (stolowski) [reply overlord/snapstate/autorefresh.go]: Sounds good.

File: overlord/snapstate/backend/link.go [hunk 2]
```diff
@@ -78,30 +79,38 @@ func hasFontConfigCache(info *snap.Info) bool {
...
 	// fontconfig is only relevant on classic and is carried by 'core' or
 	// 'snapd' snaps
 	// for non-core snaps, fontconfig cache needs to be updated before the
 	// snap applications are runnable
 	if release.OnClassic && !hasFontConfigCache(info) {
-		// XXX: does this need cleaning up? (afaict no)
-		if err := updateFontconfigCaches(); err != nil {
-			logger.Noticef("cannot update fontconfig cache: %v", err)
-		}
+		timings.Run(tm, "update-fc-cache[non-core]", "update font config cache", func(timings.Measurer) {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/snapstate/backend/link.go]: Indeed, good point.

File: overlord/patch/patch5.go [hunk 3]
```diff
@@ -63,7 +64,7 @@ func patch5(st *state.State) error {
 			continue
 		}
 
-		err = wrappers.StopServices(svcs, snap.StopReasonRefresh, log)
+		err = wrappers.StopServices(svcs, snap.StopReasonRefresh, log, nil)
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/patch/patch5.go]: Great catch, thank you!

File: overlord/snapstate/handlers.go [hunk 4]
```diff
@@ -717,7 +756,11 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 	}
 
 	pb := NewTaskProgressAdapterUnlocked(t)
-	if copyDataErr := m.backend.CopySnapData(newInfo, oldInfo, pb); copyDataErr != nil {
+	var copyDataErr error
+	timings.Run(perfTimings, "copy-snap-data", fmt.Sprintf("copy data of snap %s", snapsup.InstanceName()), func(timings.Measurer) {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply overlord/snapstate/handlers.go]: Fair point, indeed, not really needed.

File: overlord/snapstate/autorefresh.go [hunk 5]
```diff
@@ -398,6 +402,10 @@ func (m *autoRefresh) launchAutoRefresh() error {
 	chg.Set("snap-names", updated)
 	chg.Set("api-data", map[string]interface{}{"snap-names": updated})
 
+	tm.Stop()
```
Comment (mvo5) [inline overlord/snapstate/autorefresh.go]: Could we put this into a `defer func() {...}` maybe to have it close to the timings.New() call?

**Pull request merged**